src/util.ts

```diff
@@ -750,6 +765,137 @@ export async function prepareXchainFromAmountCalculation({
 	}
 }
 
+async function routeWithSlippage({
+	slippagePercentage,
+	fromToken,
+	toToken,
+	targetAmount,
+	squidRouterUrl,
+	fromAddress,
+	toAddress,
+	fromTokenPrice,
+	toTokenPrice,
+}: {
+	slippagePercentage: number
+	fromToken: TokenData
+	toToken: TokenData
+	targetAmount: string
+	squidRouterUrl: string
+	fromAddress: string
+	toAddress: string
+	fromTokenPrice: number
+	toTokenPrice: number
+}): Promise<{
+	estimatedFromAmount: string
+	weiFromAmount: ethers.BigNumber
+	routeResult: interfaces.ISquidRoute
+}> {
+	const estimatedFromAmount = await prepareXchainFromAmountCalculation({
+		fromToken,
+		toAmount: targetAmount,
+		toToken,
+		slippagePercentage,
+		fromTokenPrice,
+		toTokenPrice,
+	})
+	console.log('estimatedFromAmount', estimatedFromAmount)
+	if (!estimatedFromAmount) {
+		throw new Error('Failed to estimate from amount')
+	}
+	const weiFromAmount = ethers.utils.parseUnits(estimatedFromAmount, fromToken.decimals)
+	const routeResult = await getSquidRoute({
+		squidRouterUrl,
+		fromChain: fromToken.chainId,
+		fromToken: fromToken.address,
+		fromAmount: weiFromAmount.toString(),
+		toChain: toToken.chainId,
+		toToken: toToken.address,
+		fromAddress,
+		toAddress,
+		enableBoost: true,
+	})
+	return { estimatedFromAmount, weiFromAmount, routeResult }
+}
+
+/**
+ * For a token pair and target amount calculates the minium from amount
+ * needed to get the target amount, and the squid route to get there
+ */
+export async function routeForTargetAmount({
+	slippagePercentage,
+	fromToken,
+	toToken,
+	targetAmount,
+	squidRouterUrl,
+	fromAddress,
+	toAddress,
+}: {
+	slippagePercentage?: number
+	fromToken: TokenData
+	toToken: TokenData
+	targetAmount: string
+	squidRouterUrl: string
+	fromAddress: string
+	toAddress: string
+}): Promise<{
+	estimatedFromAmount: string
+	weiFromAmount: ethers.BigNumber
+	routeResult: interfaces.ISquidRoute
+	finalSlippage: number
+}> {
+	const fromTokenPrice = await getTokenPrice({
+		chainId: fromToken.chainId,
+		tokenAddress: fromToken.address,
+	})
+	const toTokenPrice = await getTokenPrice({
+		chainId: toToken.chainId,
+		tokenAddress: toToken.address,
+	})
```

Comment:
Recommendation: Why not promise.all here? We used it above

Reply:
done